### ISSUE TITLE
Missing required package silently ignored during prepare

### ISSUE THREAD

**Issue by sopos** [psss/tmt#516]
The required beakerlib library is expected to be installed from the yum repo. The package is unavailable; the tmt, however, continuous anyway causing the test to fail due to unsatisfied _require_.

```
...
        how: install
        summary: Install required packages
        name: requires
        order: 70
            Check if sudo is necessary.
            Execute command 'whoami' on guest 'sopos-rhel7-brq.usersys.redhat.com'.
            Run command 'ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null root@sopos-rhel7-brq.usersys.redhat.com whoami'.
            err: Warning: Permanently added 'sopos-rhel7-brq.usersys.redhat.com,10.43.21.131' (ECDSA) to the list of known hosts.
            out: root
            Check if dnf is available.
            Execute command 'rpm -q dnf' on guest 'sopos-rhel7-brq.usersys.redhat.com'.
            Run command 'ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null root@sopos-rhel7-brq.usersys.redhat.com rpm -q dnf'.
            err: Warning: Permanently added 'sopos-rhel7-brq.usersys.redhat.com,10.43.21.131' (ECDSA) to the list of known hosts.
            out: package dnf is not installed
            Command returned '1'.
            Using 'yum' for all package operations.
        package: 2 packages requested
            beakerlib
            library(realmd/basic)
            Execute command 'yum --cacheonly install -y "beakerlib" "library(realmd/basic)" || yum install -y "beakerlib" "library(realmd/basic)"' on guest 'sopos-rhel7-brq.usersys.redhat.com'.                                                                                                                             
            Run command 'ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null root@sopos-rhel7-brq.usersys.redhat.com yum --cacheonly install -y "beakerlib" "library(realmd/basic)" || yum install -y "beakerlib" "library(realmd/basic)"'.
            err: Warning: Permanently added 'sopos-rhel7-brq.usersys.redhat.com,10.43.21.131' (ECDSA) to the list of known hosts.
            out: Loaded plugins: langpacks, product-id, search-disabled-repos, subscription-
            out:               : manager
            out: 
            out: This system is not registered with an entitlement server. You can use subscription-manager to register.
            out: 
            out: Package matching beakerlib-1.18-7.el7bkr.noarch already installed. Checking for update.
            out: No package library(realmd/basic) available.
            out: Nothing to do
        how: install
        summary: Install recommended packages
        name: recommends
        order: 71
...
```

**Comment by lukaszachy** [comment]
Thank you, setting 'must' and 'bug'. Per documentation: ` If the package cannot be installed test execution should result in an error.`

**Comment by sopos** [comment]
this is also easily to reproducible using just `require: blabla`

**Comment by psss** [comment]
Hm, interesting, I see a completely different result:

```
> tmt run -avd provision -h container
Using tree '/tmp/bla'.
/var/tmp/tmt/run-459
Run data not found.
Enabled steps: discover, prepare, execute, finish, provision and report
Found 1 plan.

/plans/example
    discover
        workdir: /var/tmp/tmt/run-459/plans/example/discover
        how: fmf
            Git root not found, using '/tmp/bla.'
        directory: /tmp/bla
            Copy '/tmp/bla' to '/var/tmp/tmt/run-459/plans/example/discover/default/tests'.
            Check metadata tree in '/var/tmp/tmt/run-459/plans/example/discover/default/tests/'.
            Repository 'https://github.com/beakerlib/realmd' not found.
        summary: 1 test selected
            /tests/example
    provision
        workdir: /var/tmp/tmt/run-459/plans/example/provision
        how: container
        image: fedora
        name: tmt-var-tmp-tmt-run-459-plans-example
        distro: Fedora 33 (Container Image)
        kernel: 5.10.7-200.fc33.x86_64
        summary: 1 guest provisioned
    prepare
        workdir: /var/tmp/tmt/run-459/plans/example/prepare
        how: install
        summary: Install required packages
        name: requires
            Using 'dnf' for all package operations.
        package: 1 package requested
            library(realmd/basic)
    finish
        workdir: /var/tmp/tmt/run-459/plans/example/finish
        container: stopped
        container: removed
        summary: 0 tasks completed

stdout (10/10 lines)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Last metadata expiration check: 6 days, 6:44:51 ago on Tue Jan 19 08:41:04 2021.
No match for argument: library(realmd/basic)
Copr repo for tmt owned by psss                 7.4 kB/s | 3.3 kB     00:00    
Copr repo for tmt owned by psss                  14 kB/s | 6.9 kB     00:00    
Fedora Modular 33 - x86_64                      1.6 kB/s |  25 kB     00:16    
Fedora Modular 33 - x86_64 - Updates             85 kB/s |  22 kB     00:00    
Fedora Modular 33 - x86_64 - Updates            1.2 MB/s | 679 kB     00:00    
Fedora 33 - x86_64 - Updates                    193 kB/s |  23 kB     00:00    
Fedora 33 - x86_64 - Updates                    3.3 MB/s | 3.9 MB     00:01    
No match for argument: library(realmd/basic)

stderr (2/2 lines)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error: Unable to find a match: library(realmd/basic)
Error: Unable to find a match: library(realmd/basic)

Failed to run command 'podman exec tmt-var-tmp-tmt-run-459-plans-example sh -c dnf --cacheonly install -y "library(realmd/basic)" || dnf install -y "library(realmd/basic)"'.

> echo $?
2
```

Test metadata look like this:

```
> cat tests/example/main.fmf 
summary: Concise summary describing what the test does
test: ./test.sh
require: library(realmd/basic)
```

@sopos, do you see any substantial difference in the config?

**Comment by sopos** [comment]
@psss , try to add some requirement which can be satisfied, like `beakerlib`

**Comment by sopos** [comment]
this is the root cause:
```
[0 root@sopos-rhel7-brq ~]# yum install beakerlib -y ; echo $?
Loaded plugins: langpacks, product-id, search-disabled-repos, subscription-manager

This system is not registered with an entitlement server. You can use subscription-manager to register.

Package matching beakerlib-1.18-7.el7bkr.noarch already installed. Checking for update.
Nothing to do
0
[0 root@sopos-rhel7-brq ~]# yum install beakerlib blabla -y ; echo $?
Loaded plugins: langpacks, product-id, search-disabled-repos, subscription-manager

This system is not registered with an entitlement server. You can use subscription-manager to register.

Package matching beakerlib-1.18-7.el7bkr.noarch already installed. Checking for update.
No package blabla available.
Nothing to do
0
[0 root@sopos-rhel7-brq ~]# yum install blabla -y ; echo $?
Loaded plugins: langpacks, product-id, search-disabled-repos, subscription-manager

This system is not registered with an entitlement server. You can use subscription-manager to register.

No package blabla available.
Error: Nothing to do
1
```

**Comment by sopos** [comment]
The requirements should be checked somehow separately, I guess

**Comment by sopos** [comment]
This might also invalidate https://github.com/psss/tmt/issues/401

**Comment by sopos** [comment]
so the whole call would be
```
rpm -q --whatprovides "${require[@]}" > /dev/null 2>&1 || yum --cacheonly -y install "${require[@]}"
rpm -q --whatprovides "${require[@]}" > /dev/null 2>&1 || yum -y install "${require[@]}"
rpm -q --whatprovides "${require[@]}" > /dev/null 2>&1
```

**Comment by psss** [comment]
Hm, so this looks like there is a bug in `yum` as `dnf` handles this scenario as expected. In the meantime we can somehow workaround that for older distros. If we can use `rpm -q --whatprovides` instead of `yum --cacheonly` to cover both package names and file paths then I think we are good.

**Comment by sopos** [comment]
Right, looks like `dnf` does not suffer by this. Anyway, the `rpm -q --whatprovides` looks to me quite bulletproof solution.

**Comment by psss** [comment]
So the `--whatprovides` option works fine for package names and file paths:

    > rpm -q --whatprovides tmt
    tmt-1.2.1-1.20210121152540371060.master.8.g52ae3d7.fc33.noarch

    > rpm -q --whatprovides /usr/bin/tmt
    tmt-1.2.1-1.20210121152540371060.master.8.g52ae3d7.fc33.noarch

Seems to be ok for the special library/abi requires as well:

    > rpm -q --whatprovides 'library(distribution/epel)'
    distribution-distribution-Library-epel-1.0-65.noarch

    > rpm -q --whatprovides 'python(abi)'
    python2.7-2.7.18-6.fc33.x86_64
    python3-3.9.1-2.fc33.x86_64

But does not seem to work with versioned python abi format:

    > rpm -q --whatprovides 'python(abi) = 3.9'
    no package provides python(abi) = 3.9

I guess that could be a problem. Thoughts?

**Comment by sopos** [comment]
Ok, you've got me. It's quite unexpected that a test would have such requirement but we should not forbid it.

**Comment by sopos** [comment]
So the only proper way I can see now is to stupidly iterate and call yum for individual requirement.
```
check() {
  ( rpm -q --whatprovides "$1" || yum install --cacheonly -y "$1" ) > /dev/null 2>&1
}
check_all() {
  local res=0
  for r in "$@"; do
    check "$r" || {
      res=1
      break
    }
  done
  [[ $res -eq 0 ]]
}
check_all "${require[@]}" || {
  yum install -y "${require[@]}" && \
  check_all "${require[@]}"
}
```

It is quite fast if there are no versioned requirements otherwise it gets slowed.

**Comment by psss** [comment]
Hm, thinking about it a bit more, if we go with `rpm -q --whatprovides` only it would nicely cover the most common use cases, while for the special require format the command would fail and result in just an extra `yum install` command which will state that everything is installed. So I guess this approach is just fine and would make the implementation much more readable.

**Comment by psss** [comment]
Ah, taking back my last comment. This would not cover the [yum bug](https://bugzilla.redhat.com/show_bug.cgi?id=1920176).

**Comment by sopos** [comment]
I do not agree. What you propose is basically a check before calling `yum`. My issue is the check after it. If `yum` fails to install everything and still reports success after all. There's a need to check it differently at the end.

**Comment by sopos** [comment]
> Ah, taking back my last comment. This would not cover the [yum bug](https://bugzilla.redhat.com/show_bug.cgi?id=1920176).

exactly

**Comment by psss** [comment]
Oh, it's getting ugly here ;-) The thing is that the code is run on the guest so the implementation you propose above would need to be synced there. Hmmm, what about this?

 * Check if everything required is installed using `rpm -q --whatprovides`
 * If not, run `yum/dnf install -y all-required`
 * Plus, when `yum` is used do `for package in requires: rpm -q --whatprovides $package || yum install $package`

In other words, do it once for all packages and then, for older distros, do an additional check one by one. Also ugly, but probably a bit easier to implement without extra code syncing.

**Comment by sopos** [comment]
It looks to me very similar to what I proposed. I admit is does not look nice as a one-liner but it is not much worse than your proposal.

**Comment by psss** [comment]
I see, agreed. Let's consider those versioned provides on older distros as a corner case. The fix #517 is ready for review.
